### PR TITLE
Fix for bug #4863, update the Proofview's env with side_effects

### PR DESCRIPTION
**Pull request by mattam82**
Partial solution to the handling of side effects in proofview. At refine time, if the sigma contains new side effects, they are synchronized with the Proofview's env.

### CODE REVIEW

**Comment by mattam82** [comment]
@aspiwack @gares Please review. It's is not the ultimate solution as other calls to Unsafe.tclEVARS  / Proofview.modify should do the same, but it seems a tid bit cleaner than putting a call to Global.env () in Env.get.

**Comment by mattam82** [comment]
Being tested here: https://ci.inria.fr/coq/view/coq-contribs/job/coq-contribs/432/console

**Comment by gares** [comment]
LGTM.

File: proofs/proofview.ml [hunk 1]
```diff
@@ -1067,6 +1067,20 @@ struct
...
+    | { Entries.eff = Entries.SEsubproof (kn, cb, eff_env) } ->
+       [ kn, cb ]
+    | { Entries.eff = Entries.SEscheme (l,_) } ->
+       List.map (fun (_,kn,cb,eff_env) ->
+           kn, cb) l
+
+  let constant_bodies_of_effects =
+    List.map_append constant_bodies_of_effect
+
+  let add_side_effects csts env =
```
Comment (aspiwack) [inline proofs/proofview.ml]: Could you add a few comments on what these three functions do?

File: kernel/safe_typing.ml [hunk 2]
```diff
@@ -222,13 +222,6 @@ let inline_private_constants_in_constr = Term_typing.inline_side_effects
 let inline_private_constants_in_definition_entry = Term_typing.inline_entry_side_effects
 let side_effects_of_private_constants x = Term_typing.uniq_seff (List.rev x)
 
-let constant_entry_of_private_constant = function
```
Comment (aspiwack) [inline kernel/safe_typing.ml]: Am I correct that this is unrelated dead code? In general, it's better to use a different PR for these, to facilitate review. To help with that, let me introduce [git worktree](https://git-scm.com/docs/git-worktree).

When a different PR is not practical, using a different commit is still desirable as it would facilitate potential reverts.

Just saying. It shouldn't block this PR.

File: proofs/proofview.ml [hunk 3]
```diff
@@ -1079,6 +1093,10 @@ struct
     let (sigma, c) = f (Evd.reset_future_goals sigma) in
     let evs = Evd.future_goals sigma in
     let evkmain = Evd.principal_future_goal sigma in
+    (** Redo the effects of sigma in the env *)
+    let privates_csts = Evd.eval_side_effects sigma in
+    let sideff = Safe_typing.side_effects_of_private_constants privates_csts in
+    let env = add_side_effects sideff env in
```
Comment (aspiwack) [inline proofs/proofview.ml]: Is there a risk of doing the same side effect twice? Would that be bad?

**Review by aspiwack — COMMENTED** [review]
I don't remember much of the side-effect thing. So my review may not be very useful. But here goes.

File: proofs/proofview.ml (hunk 1 above)
Comment (mattam82) [inline proofs/proofview.ml]: Yes, will do

File: proofs/proofview.ml (hunk 3 above)
Comment (mattam82) [inline proofs/proofview.ml]: Yes, and no. It will just replace the definition of a global constant with the same entry (i.e., the identity).

File: kernel/safe_typing.ml (hunk 2 above)
Comment (mattam82) [inline kernel/safe_typing.ml]: Yes I agree. It's just dead code I spotted when working on this.

File: proofs/proofview.ml (hunk 3 above)
Comment (mattam82) [inline proofs/proofview.ml]: I'm updating the PR to do this correctly, checking if the global was already declared or not and avoid the intermediary list datastructure.

**Comment by mattam82** [comment]
Updated, testing the changed code here:
https://ci.inria.fr/coq/view/coq-contribs/job/coq-contribs/465/console
